Cbrelease 3.0.2 rating review features

wilkysingh-fortum opened the pull request:
Add rating code is implemented.

wilkysingh-fortum committed:
add Rating api code implementation

wilkysingh-fortum committed:
Merge branch 'cbrelease-3.0.2' of https://github.com/sunbird-cb/sunbird-cb-ext into cbrelease-3.0.2-rating-review-features

wilkysingh-fortum committed:
Redactoring done for messageEvent and upsert ratings

wilkysingh-fortum committed:
Merge branch 'cbrelease-3.0.2' of https://github.com/sunbird-cb/sunbird-cb-ext into cbrelease-3.0.2-rating-review-features

mathewjpallan commented on src/main/java/org/sunbird/ratings/controller/RatingsController.java:
```diff
@@ -0,0 +1,48 @@
+    // ----------------- Public APIs --------------------
+
+    @PostMapping("/ratings/v1/upsert")
+    public ResponseEntity<?> upsert(@Valid @RequestBody RequestRating requestRatingBody) throws Exception {
```
Should we just call this the update API?

wilkysingh-fortum replied:
We are doing both addition and update in this so named it upsert to give more clarity about its working.

mathewjpallan commented on src/main/java/org/sunbird/ratings/controller/RatingsController.java:
```diff
@@ -0,0 +1,48 @@
+//        return new ResponseEntity<>(ratingService.getUsers(activity_Id, activity_Type, userId), HttpStatus.OK);
+//    }
+
+//    @PostMapping("/ratings/v1/update")
```
This API is not required anymore

wilkysingh-fortum replied:
Deleted.

mathewjpallan commented on src/main/java/org/sunbird/ratings/exception/ValidateRatingException.java:
```diff
@@ -0,0 +1,24 @@
+
+@ResponseStatus(value = HttpStatus.BAD_REQUEST)
+@ResponseBody
+public class ValidateRatingException extends RuntimeException {
```
Can we just have this class as ValidationException and introduce a list of error codes and have a list of messages also. Any API wishing to send a validation error should be able to reuse this. The reason for asking list of messages is because an API may want to perform all validations and give out a list of validation issues

wilkysingh-fortum replied:
These seems good add on.Will do it

mathewjpallan commented on src/main/java/org/sunbird/ratings/model/RatingMessage.java:
```diff
@@ -0,0 +1,74 @@
+package org.sunbird.ratings.model;
+
+public class RatingMessage {
```
Can we not reuse one bean class instead of RatingMessage and RatingRequest? Both can have the same params right?

wilkysingh-fortum replied:
no there are couple of different params for example we are not maintaining version etc. and also have extra previous values and updated values attributes

mathewjpallan commented on src/main/java/org/sunbird/ratings/service/RatingService.java:
```diff
@@ -0,0 +1,17 @@
+
+    public SBApiResponse getRatingSummary(String activity_Id, String activity_Type) throws Exception; //To do
+
+    public SBApiResponse upsert(RequestRating requestRating) throws Exception;
```
We should never throw Exception which is a catchall. We need not even declare a throws clause as the custom exception thrown is a runtime exception.

wilkysingh-fortum replied:
done

mathewjpallan commented on src/main/java/org/sunbird/ratings/model/UpdatedValues.java:
```diff
@@ -0,0 +1,32 @@
+package org.sunbird.ratings.model;
+
+public class UpdatedValues {
```
Is there a neat way to include this in the Rating bean class itself? Having a separate POJO does not seem like a good idea

wilkysingh-fortum replied:
Done

mathewjpallan commented on src/main/java/org/sunbird/ratings/service/RatingServiceImpl.java:
```diff
@@ -0,0 +1,174 @@
+            request.put(Constants.REVIEW, requestRating.getReview());
+            request.put(Constants.UPDATED_ON, timeBasedUuid);
+
+            cassandraOperation.insertRecord(Constants.KEYSPACE_SUNBIRD, Constants.TABLE_RATINGS, request);
```
is the insertRecord function doing an upsert if the row is already present?

mathewjpallan replied:
or does the updateRecord add the record if it does not exist? in that case you do not need 2 sep calls to insert and update.

wilkysingh-fortum replied:
No It doesn't validate whether the row already exists.

mathewjpallan commented on src/main/java/org/sunbird/ratings/service/RatingServiceImpl.java:
```diff
@@ -0,0 +1,174 @@
+        } catch (ValidateRatingException ex) {
+            String errMsg = "Exception is : ";
+            return processExceptionBody(response, ex, errMsg, HttpStatus.BAD_REQUEST);
+        } catch (Exception ex) {
```
We should not have a catch block for Exception which is very broad.

wilkysingh-fortum replied:
as discussed leaving it as it can catch other cassandra or unexpected sorts of errors

mathewjpallan commented on src/main/java/org/sunbird/ratings/service/RatingServiceImpl.java:
```diff
@@ -0,0 +1,174 @@
+                cassandraOperation.updateRecord(Constants.KEYSPACE_SUNBIRD, Constants.TABLE_RATINGS, updateRequest,
+                        request);
+
+                msg.setAction("ratingUpdate");
```
We should have a constructor that takes in the basic types so that we can avoid so many setter calls

wilkysingh-fortum replied:
done

mathewjpallan commented on src/main/java/org/sunbird/ratings/service/RatingServiceImpl.java:
```diff
@@ -0,0 +1,174 @@
+    CassandraOperation cassandraOperation;
+
+    @Autowired
+    Producer producer;
```
Rename variable as kafka producer

wilkysingh-fortum replied:
ok

wilkysingh-fortum replied:
Done

mathewjpallan commented on src/main/java/org/sunbird/ratings/service/RatingServiceImpl.java:
```diff
@@ -0,0 +1,174 @@
+                msg.setPrevValues(prevValues);
+                msg.setUpdatedValues(updatedValues);
+
+                producer.push(Constants.RATINGS_UPDATE_EVENT, msg);
```
The topic name should be read from env variables and not from constants. This is required because the topic names would be namespaced with env names like dev.topicname/stage.topicname etc

wilkysingh-fortum replied:
done

mathewjpallan commented on src/main/java/org/sunbird/ratings/service/RatingServiceImpl.java:
```diff
@@ -0,0 +1,174 @@
+            request.put(Constants.RATING, requestRating.getRating());
+            request.put(Constants.REVIEW, requestRating.getReview());
+            request.put(Constants.UPDATED_ON, timeBasedUuid);
+
```
The event needs to be sent to Kafka even if prevValues do not exist.

wilkysingh-fortum replied:
done

wilkysingh-fortum replied:
Done

mathewjpallan reviewed (commented):
Please address these comments

wilkysingh-fortum committed:
Read rating api added and refactoring done as per comment on PR

mathewjpallan committed:
Minor edits
- Changed the order of the path params as activity_type should come first
- Changed the response code for all scenarios to 200
- Removed lint errors
